use mrcfile library for MRC/CCP4 file parsing

orbeckst opened the pull request:
- fix #83
- use mrcfile https://github.com/ccpem/mrcfile (light-weight dependency) for file parsing
- new mrc module; functionality replaces CCP4 module
- keep CCP4 module for one release cycle
- add tests

pep8speaks commented:
Hello @orbeckst! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`gridData/mrc.py`](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py):

> [Line 66:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py#L66): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 74:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 75:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 77:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 101:1](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/mrc.py#L101): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`gridData/tests/test_mrc.py`](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py):

> [Line 13:1](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L13): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 17:1](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L17): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 29:1](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L29): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 31:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L31): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 32:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L32): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 33:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L33): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 35:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L35): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 36:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L36): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 37:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L37): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 39:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L39): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 40:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L40): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 41:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L41): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 45:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L45): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 46:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L46): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 47:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L47): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 48:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L48): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 49:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L49): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 50:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L50): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 54:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L54): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 55:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L55): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 56:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L56): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 59:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L59): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 60:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L60): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 61:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L61): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 62:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L62): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 63:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L63): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 64:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L64): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 67:5](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L67): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 67:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 69:25](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L69): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 71:80](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L71): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 78:1](https://github.com/MDAnalysis/GridDataFormats/blob/fbdd4914988e16ecc9b46bef92bb2dbe6ed57230/gridData/tests/test_mrc.py#L78): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

orbeckst commented:
For reviewers: I intend to ignore the PEP8 formatting warnings... if you feel strongly about it then I suggest we run everything through black from here on out. I don't have the time to spare to manually prettify code that started out pretty ugly anyway.

orbeckst commented:
@MDAnalysis/coredevs if you want to have a quick look then that would be much appreciated. If you think it's good enough then you can approve it — I am mainly looking for a second pair of eyes to avoid really stupid mistakes.

If nobody has time I'll merged it by the end of the week latest in order to get a release out.

IAlibay commented:
> If nobody has time I'll merge it by the end of the week latest in order to get a release out.

I've penciled time in to review on Thursday if that's ok? (assuming no one else gets to review first)

tluchko commented:
@orbeckst, thank you for implementing this.  Doing so has been on my to-do list for a long time but never managed to get to it.

The code looks good to me but I have two questions about functionality.

1. Why are non-orthorhombic unit cells not supported?  It seems they are supported for the DX format, as I get no complaints when I load such a file.
2. Can we add support for writing MRC files?  I played around with this and I think it will be straightforward.  However, I don't know how to recover the unit cell angles for non-orthorhombic from a `Grid` object.

I can add writing MRC files for orthorhombic cells, if you like. Once I understand the issues for non-orthorhombic cells, I should be able to add read/write support for these as well.

IAlibay commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+        if filename is not None:
+            self.filename = filename
+        with mrcfile.open(filename) as mrc:
+            if not mrc.is_volume():                           #pragma: no cover
```
is calling validate worth it too? (not sure how much of an issue a non-valid mrc file is)

orbeckst replied:
You might want to read a file with an iffy header and just get the data. Not sure, TBH.

orbeckst replied:
No implementing for now because I don't understand the implication/standard use.

IAlibay commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+            if not mrc.is_volume():                           #pragma: no cover
+                raise ValueError(
+                    "MRC file {} is not a volumetric density.".format(filename))
+            self.header = h = mrc.header.copy()
```
any difference between header and extended_header here?

orbeckst replied:
don't know — experts??

tluchko replied:
From https://www.ccpem.ac.uk/mrc_format/mrc2014.php, 

> The extended header is now used by different software to hold various additional metadata instead, as indicated by the EXTTYP tag. 

The `EXTTYP` flag takes seven different values. You could make a copy incase the user wants it.  Note that `self.header` does not seem to be accessible from `Grid` or, at least, it is not obvious to me how to access it.

orbeckst replied:
Do you want to have access to `header` ?

tluchko replied:
I don't have a personal-use case in mind.  But, if it isn't difficult, it could be beneficial in cases we haven't thought of yet. Not everything in the `header` ends up in a `Grid` object.

orbeckst replied:
I am adding the header as undocumented `Grid._mrc_header`. If it is used then we can think of a better way to keep format-specific data.

IAlibay commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+        if filename is not None:
+            self.filename = filename
+        with mrcfile.open(filename) as mrc:
+            if not mrc.is_volume():                           #pragma: no cover
```
Should there be an attempt to use `set_volume()` to change from an image stack to a volume?

orbeckst replied:
I don't want to deal with imagestacks — no experience and no idea how to test. I would leave this for someone who wants the feature.

IAlibay commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+                                 "supported, not "
+                                 "alpha={0}, beta={1}, gamma={2}".format(
+                                     h.cellb.alpha, h.cellb.beta, h.cellb.gamma))
+            # mrc.data[z, y, x] indexed: convert to x,y,z as used in GridDataFormats
```
Do we need to do further manipulations here?

```
Note that the MRC format allows the data axes to be swapped using the header’s mapc, mapr and maps fields. This library does not attempt to swap the axes and simply assigns the columns to X, rows to Y and sections to Z. (The data array is indexed in C style, so data values can be accessed using mrc.data[z][y][x].) In general, EM data is written using the default axes, but crystallographic data files might use swapped axes in certain space groups – if this might matter to you, you should check the mapc, mapr and maps fields after opening the file and consider transposing the data array if necessary.
```

tluchko replied:
I agree.  If the file has `mapc, mapr, maps = 3, 2, 1`, then the data should already be in C-order and `np.transpose()` would be unnecessary.  `mapc, mapr, maps = 1, 2, 3` is Fortran-order.  It might be best to check for these two cases and disallow the rest. `np.transpose()` should be able to handle the other cases but it they are probably rare.

orbeckst replied:
Good point, this needs to be fixed then. Perhaps that fixes #76, too.

Code welcome!!!

orbeckst replied:
Ultimately, this and calculating the origin correctly did fix #76

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,97 @@
+])
+def test_ccp4_read_header(ccp4data, name, value):
+    if type(value) is float:
+        assert_almost_equal(ccp4data.header[name], value, decimal=6)
```
I think the convention is now to use `assert_allclose` instead of `assert_almost_equal`.

I fee like there _should_ be a nicer way to do this comparison, but at the end of the day it works so there's no point trying to find the prettiest answer to this problem.

orbeckst replied:
YEs, can be changed, of course. I copied & pasted what we had before.

orbeckst replied:
kept it,,, need sleep

IAlibay reviewed (changes requested):
Everything but the one unnecessary import is a question.

orbeckst commented:
Thanks for review and comments. This week I definitely won't have time to look at anything, sorry.

orbeckst commented on gridData/tests/datafiles/__init__.py:
```diff
@@ -10,6 +10,8 @@
 # (see issue #57)
 CCP4_1JZV = resource_filename(__name__, '1jzv.ccp4')
+# from https://github.com/ccpem/mrcfile/blob/master/tests/test_data/EMD-3001.map.bz2
+MRC_EMD3001 = resource_filename(__name__, 'EMD-3001.map.bz2')
```
BSD3 is fine inside GPL. We can add a note but it's more likely that the data does not fall under the license anyway.

orbeckst commented:
@tluchko can we try to get the simple reading code in and then add the enhancements, namely writing?

For the non-orthorhombic cells I can't quite remember why I restricted it to orthorhombic. There might be something in the Grid class that's not working with triclinic cells. I'd try to get this PR done with the limitation and then raise a new issue for triclinic cells.

tluchko reviewed (commented):
I suggested some code to deal with the different potential orderings of `mrc.data`.

orbeckst commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+                                 "alpha={0}, beta={1}, gamma={2}".format(
+                                     h.cellb.alpha, h.cellb.beta, h.cellb.gamma))
+            # mrc.data[z, y, x] indexed: convert to x,y,z as used in GridDataFormats
+            self.array = np.transpose(mrc.data)
```
Thank you. I am adding your code, we'll still need a test.

Sorry to be slow, I have a ton of other things to attend to — beginning of semester is always stressful.

orbeckst replied:
todo here
- [ ] test
- [ ] add @tluchko to CHANGELOG

orbeckst commented on gridData/core.py:
```diff
@@ -123,7 +123,8 @@ def __init__(self, grid=None, edges=None, origin=None, delta=None,
         }
         self._loaders = {
-            'CCP4': self._load_cpp4,
+            'CCP4': self._load_mrc,
```
I really only see this through the lens of the tests and if it doesn't break the tests then it didn't change anything. That's probably not true, though. There should definitely be a versionchanged entry to make sure users have a fighting chance to notice.

orbeckst replied:
added versionchanged

orbeckst commented on gridData/CCP4.py:
```diff
@@ -160,6 +163,9 @@ class CCP4(object):
        * index ordering besides standard column-major and row-major
        * non-standard fields, such any in filed in future use block
+
+    .. deprecated:: 0.7.0
```
Need to check. We should have a RTD build in the PR – needs to be added to CI #104 .

orbeckst replied:
done

orbeckst commented:
The code with sorted mapc/r/s  transposed axes does _not_ fix #76 . In ChimeraX
* Opened **1cbs.ccp4** as 2, **grid size 81,70,96**, pixel 0.601,0.595,0.588, (= y, x, z)
* Opened **1cbs.dx** as 3, **grid size 70,96,81,** pixel 0.601,0.595,0.588 (=x, z, y)

1cbs.ccp4 :
* mapc, mapr, maps: 2, 1, 3.
* nx, ny, nz: 70, 81, 96
* mrc.data.shape = (96, 81, 70) (automatically reordered into z, x, y)

This looks as if mrcfile does indeed *always* reorder the data into z, x, y order.

orbeckst commented on gridData/mrc.py:
```diff
@@ -0,0 +1,136 @@
+from __future__ import absolute_import
+from six.moves import range
+
+import warnings
```
removed

orbeckst commented:
@tluchko and @IAlibay I addressed most of your comments and after a long time thinking about what mrcfile does and what the mapc/r/s parameter mean I managed to get the conversion of MRC data to Grid to work (and fixed the super-irritating #76 along the way). This should be good for a final review.

If you can find a nicer way to write the double-transposition
```python
            axes_order = np.hstack([h.mapc, h.mapr, h.maps])
            axes_c_order = np.argsort(axes_order)
            self.array = np.transpose(np.transpose(mrc.data), axes=axes_c_order)
```
then let me know. However, just reversing `axes_c_order` does not work as far as I can tell... there must be a way to permute the `axes_c_order` directly but I am too tired to figure it out. (The first transpose is always (210) (rearrange zyx to xyz), the second one depends on the mapc/r/s values and is in the test (102) (x is really y, y is really x, and z is z). The end result needs to be (201) ((zyx) -> (yxz)).

orbeckst commented:
Honestly, I don't know why reversing `axes_order_c` did not work for me because (102) reversed is (201) and that's also what you get when applying (210) to (102). But 
```python
self.array = np.transpose(mrc.data, axes=axes_c_order[::-1])
```
gives the wrong shape.

tluchko commented on gridData/mrc.py:
```diff
@@ -107,14 +105,24 @@ def read(self, filename):
+            # All other "xyz" quantitities are also reordered.
+            axes_order = np.hstack([h.mapc, h.mapr, h.maps])
+            axes_c_order = np.argsort(axes_order)
+            self.array = np.transpose(np.transpose(mrc.data), axes=axes_c_order)
+            self.delta = np.diag(
+                np.array([mrc.voxel_size.x, mrc.voxel_size.y, mrc.voxel_size.z])[axes_c_order])
+            # the grid is shifted to the CCP4 origin by offset
+            # (assume orthorhombic)
+            offsets = np.hstack([h.nxstart, h.nystart, h.nzstart])[axes_c_order] * np.diag(self.delta)
+            # GridData origin is centre of cell at x=col=0, y=row=0 z=seg=0
+            self.origin = np.hstack([h.origin.x, h.origin.y, h.origin.z])[axes_c_order] + offsets
```
The code below works with only one transpose.  Because `mrcfile` transposes the data, we need to reverse the order of `axes_order` before sorting. Note that `voxelsize`  and `origin`  refer to the x, y, and z real space axes and should not be reordered.  However, `nxstart`, `nystart` and `nzstart` should be reordered using `axes_c_order`. I'm not sure why, but I had to do this to match what ChimeraX and VMD are reporting for the example file in #76.
```suggestion
            transpose_order = np.argsort(axes_order[::-1])
            self.array = np.transpose(mrc.data, axes=transpose_order)
            self.delta = np.diag(np.array([mrc.voxel_size.x, mrc.voxel_size.y, mrc.voxel_size.z]))
            # the grid is shifted to the MRC origin by offset
            # (assume orthorhombic)
            offsets = np.hstack([h.nxstart, h.nystart, h.nzstart])[axes_c_order] * np.diag(self.delta)
            # GridData origin is centre of cell at x=col=0, y=row=0 z=seg=0
            self.origin = np.hstack([h.origin.x, h.origin.y, h.origin.z]) + offsets
```

orbeckst replied:
Thanks, I am going with your suggestion then. I am not an expert on the MRC format and from the description it wasn't clear to me what corresponds to which direction. 

I didn't include the data from #76 as a new test because it looked as if the included test file 1jzv.ccp4 for CCP4 would be sufficient (it also contains reordered axes). However, if you think that there's something in 1cbs.ccp4 that would make for a good test case then we can include it (especially if 1CBS showed you problems that were not apparent with 1jzv.ccp4).

tluchko replied:
It looks like 1cbs.ccp4 doesn't add anything that isn't in 1jzv.ccp4 already. Both have the same use of `mapc`/`r`/`s`, `nstart` and `origin`. 

Looking back, `nx`, `ny`, `nz`, `nstart` and the data are all in reference to the grid, so the transpose/reorder specified by `mapc`/`r`/`s` must be applied.  OTOH, `mx`, `my`, `mz`, `cella`, `cellb` and `origin` refer to the real-space dimensions and do not require reordering. It would be nice to have additional test cases but I think these properties are handled correctly now.

orbeckst replied:
Thanks for double checking, much appreciated!

tluchko reviewed (changes requested):
I suggested some revised code to avoid the double transpose of the data and match the delta and origin reported for test files in ChimeraX and VMD.

orbeckst commented:
I rewrote the history so that this can be merged as is.

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+import pytest
+
+import numpy as np
+from numpy.testing import (assert_almost_equal,
```
```suggestion
from numpy.testing import (assert_allclose,
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+])
+def test_ccp4_read_header(ccp4data, name, value):
+    if type(value) is float:
+        assert_almost_equal(ccp4data.header[name], value, decimal=6)
```
```suggestion
        assert_allclose(ccp4data.header[name], value, rtol=1e-06)
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+    assert_equal(ccp4data.shape, (76, 96, 70))
+
+def test_delta(ccp4data):
+    assert_almost_equal(ccp4data.delta, np.array(
```
```suggestion
    assert_allclose(ccp4data.delta, np.array(
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+    #
+    # (visual comparison of CCP4 and DX file in ChimeraX at same
+    # level shows full agreement)
+    assert_almost_equal(ccp4data.origin, [-12.5404758,  -2.1809523,  57.151876 ])
```
```suggestion
    assert_allclose(ccp4data.origin, [-12.5404758,  -2.1809523,  57.151876 ])
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+        assert grid._mrc_header == ccp4data.header
+
+    def test_delta(self, grid, ccp4data):
+        assert_almost_equal(grid.delta, np.diag(ccp4data.delta))
```
```suggestion
        assert_allclose(grid.delta, np.diag(ccp4data.delta))
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+        assert_almost_equal(grid.delta, np.diag(ccp4data.delta))
+
+    def test_origin(self, grid, ccp4data):
+        assert_almost_equal(grid.origin, ccp4data.origin)
```
```suggestion
        assert_allclose(grid.origin, ccp4data.origin)
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+        assert_almost_equal(grid.origin, ccp4data.origin)
+
+    def test_data(self, grid, ccp4data):
+        assert_almost_equal(grid.grid, ccp4data.array)
```
```suggestion
        assert_allclose(grid.grid, ccp4data.array)
```

IAlibay commented on gridData/tests/test_mrc.py:
```diff
@@ -0,0 +1,137 @@
+    POINTS = 192
+    assert_equal(g.grid.flat, np.arange(1, POINTS+1))
+    assert_equal(g.grid.size, POINTS)
+    assert_almost_equal(g.delta, [3./4, .5, 2./3])
```
```suggestion
    assert_allclose(g.delta, [3./4, .5, 2./3])
```

IAlibay reviewed (approved):
I'll be honest here I don't really have enough time to fully understand what is being done with the MRC file transpose. From a quick look it seems to make sense and being no MRC expert I am happy to default to yours and @tluchko's expertise here.

I'm not sure what was meant by https://github.com/MDAnalysis/GridDataFormats/pull/100/files#r810475496 but I've added a couple of suggested changes for assert_almost_equal -> assert_allclose if it's still of interest to make the switch.

orbeckst commented on CHANGELOG:
```diff
@@ -13,6 +13,25 @@ The rules for this file:
   * accompany each entry with github issue/PR number (Issue #xyz)
 
 ------------------------------------------------------------------------------
+??/??/2022 orbeckst, tluchko
```
```suggestion
??/??/2022 orbeckst, tluchko, IAlibay
```

orbeckst commented:
Thank you @tluchko and @IAlibay !

orbeckst commented:
Following up on https://github.com/MDAnalysis/GridDataFormats/pull/100#issuecomment-1009487765, @tluchko please open issues for
* writing MRC files
* handling of triclinic boxes

Thanks!

tluchko commented:
Thank you @orbeckst!
